docs(1367): tidy up for tutorials (#1401)
MetaMask/metamask-docs: snaps/learn/tutorials/transaction-insights.md

---
description: Create a Snap that provides transaction insights.
sidebar_position: 2

> toc_max_heading_level: 4

Author: ```suggestion
```

[`@metamask/create-snap`](https://github.com/MetaMask/snaps/tree/main/packages/create-snap)
starter kit by running:

<Tabs>

>   <TabItem value="yarn" label="yarn" default>

Reviewer: ```suggestion
  <TabItem value="Yarn" label="Yarn" default>
```

```bash
yarn create @metamask/snap transaction-insights-snap
```
  </TabItem>
  <TabItem value="npx" label="npx" >

```bash
npx @metamask/create-snap transaction-insights-snap
```

  </TabItem>
    <TabItem value="npm" label="npm" >

```bash
npm create @metamask/snap transaction-insights-snap
```
  </TabItem>
</Tabs>

Next, `cd` into the `transaction-insights-snap` project directory and run:

```bash
yarn install
```
This initializes your development environment with the required dependencies. 

<details>
  <summary>Did you get a warning?</summary>
  <div>
You may get a warning such as: 

```bash
@lavamoat/allow-scripts has detected dependencies without configuration. explicit configuration required.
run "allow-scripts auto" to automatically populate the configuration.
```

You can resolve the issue by running: 

```bash 
yarn run allow-scripts auto
``` 
 </div>
</details>

>  
> 

Reviewer: ```suggestion
```

No multiple blank lines

the value `"images/gas.svg"` to point to your new icon:

```json title="snap.manifest.json"
"location": {
  "npm": {
    "filePath": "dist/bundle.js",
    "iconPath": "images/gas.svg",
    "packageName": "snap",
    "registry": "https://registry.npmjs.org/"
  }
}
```

Open `packages/snap/package.json` in a text editor. Edit the `files` array and reference the 
`images/` folder:

```json title="package.json"
"files": [
  "dist/",
  "images/",
  "snap.manifest.json"
],
```

#### 2.2. Update your Snap's wallet prompt

It's important that user's understand what they're agreeing to when responding to wallet prompts. The 
prompt uses the `proposedName` of the Snap, currently "Typescript Example" in our template.

Open `packages/snap/snap.manifest.json` in a text editor.
Edit the `"proposedName"` property within the metadata to provide a functional name such as "Get Fee %":

```json title="snap.manifest.json"
{
  "version": "0.1.0",
  "description": "An example Snap written in TypeScript.",
  "proposedName": "Get Fee %",
  "repository": {
    "type": "git",
    "url": "https://github.com/MetaMask/template-snap-monorepo.git"

>   },

Author: I had an idea for this -- would require tabbing between fragment and full file to clarify

>   },

Author: @alexandratran PTAL at the "fragments" vs "full page" example --> I agree that it can be confusing to look at fragments that don't looks as though they open/close properly (even though, in context, they do) WDYT does this help?

>   },

Reviewer: OK, I see this comment here now. I think I prefer just the fragments. I find the tabs with the full file a bit confusing. If necessary, you can add an ellipses "..." before the closing bracket to denote additional text.

```

#### 2.3. Update your Snap's button

Open `packages/site/src/components/Buttons.tsx` in a text editor.
Edit the Button property to provide a functional name such as "Estimate Gas ":

```typescript title="Buttons.tsx"
export const SendHelloButton = (props: ComponentProps<typeof Button>) => {
  return <Button {...props}>Inoperable</Button>;
};
````

> These three updates are the minimum required to ensure that each user interaction with your Snap is well 

Author: This is important! When you are a numpty like me a non-functional button is something you press, wonder why it throws error messages at you and then forget that this tutorial won't be using a button and try and fix da button's function.

<details>
  <summary>Icon attribution</summary>
  <div>
This is a free icon, "Gas" by Mello from the [Noun Project](https://thenounproject.com/browse/icons/term/gas/).
 </div>
</details>

Your file structure should look like this:

```text
transaction-insights-snap/
├─ packages/
│  ├─ site/
|  |  |- src/
|  |  |  |- App.tsx
|  |  ├─ package.json
|  |  |- ...(React app content)
|  |
│  ├─ snap/
|  |  ├─ images/
|  |  |  |- gas.svg
|  |  ├─ src/
|  |  |  |- index.test.ts
|  |  |  |- index.ts
|  |  ├─ snap.manifest.json
|  |  ├─ package.json
|  |  |- ... (Snap content)
├─ package.json
├─ ... (other stuff)
```

Open `packages/snap/snap.manifest.json` in a text editor. This file contains the main configuration 
details for your Snap. Edit the `npm` object, within the `location` object, and add `iconPath` with 
the value `"images/gas.svg"` to point to your new icon:

<Tabs>
  <TabItem value="Line updates" label="Update snap.manifest.json" default>

```json title="snap.manifest.json"
"location": {
  "npm": {
    "filePath": "dist/bundle.js",
    "iconPath": "images/gas.svg",
    "packageName": "snap",
    "registry": "https://registry.npmjs.org/"
  }
```
</TabItem>

> <TabItem value="Full page" label="Example" default>

Reviewer: I don't understand why this example tab is necessary? Can we remove it and make this step the same as in the "Estimate gas" tutorial?

```json title="snap.manifest.json"
{
  "version": "0.1.0",
  "description": "An example Snap written in TypeScript.",
  "proposedName": "TypeScript Example",
  "repository": {
    "type": "git",
    "url": "https://github.com/MetaMask/template-snap-monorepo.git"
  },
  "source": {
    "shasum": "TwVBeJ/PL4uARBEFZmgJukYd3XWyTbilkOPhtdg+GvU=",
    "location": {
      "npm": {
        "filePath": "dist/bundle.js",
        "iconPath": "images/gas.svg",
        "packageName": "snap",
        "registry": "https://registry.npmjs.org/"
      }
    }
  },
   "initialPermissions": {
    "snap_dialog": {},
    "endowment:rpc": {
      "dapps": true,
      "snaps": false
    },
    "endowment:network-access": {}
  },
  "manifestVersion": "0.1"
}
```
</TabItem>
</Tabs>

Open `packages/snap/package.json` in a text editor. Edit the `files` array and reference the 
`images/` folder:

<Tabs>
  <TabItem value="Line updates" label="Update package.json" default>

```json title="package.json"
"files": [
  "dist/",
  "images/",
  "snap.manifest.json"
],
```

</TabItem>

> <TabItem value="Full page" label="Example" default>

Reviewer: Same comment as above, I think we can remove this example tab.

Author: Sure -- would love some user testing to know how low our entry-level is, but will elipse code fragments instead

    "eslint-plugin-n": "^15.7.0",
    "eslint-plugin-prettier": "^4.2.1",
    "eslint-plugin-promise": "^6.1.1",
    "jest": "^29.5.0",
    "prettier": "^2.7.1",
    "prettier-plugin-packagejson": "^2.2.11",
    "rimraf": "^3.0.2",
    "ts-jest": "^29.1.0",
    "typescript": "^4.7.4"
  },
  "packageManager": "yarn@3.2.1",
  "engines": {
    "node": ">=18.6.0"
  },
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  }
}

```
</TabItem>
</Tabs>

#### 2.2. Update your Snap's name

Update your Snap's name in MetaMask.
MetaMask uses the `proposedName` of the Snap, currently "TypeScript Example" in the template.

Open `packages/snap/snap.manifest.json` in a text editor.
Edit the `"proposedName"` property within the metadata to provide a functional name such as "Transaction insights":

<Tabs>
  <TabItem value="Line updates" label="Update snap.manifest.json" default>

```json title="snap.manifest.json"
{
  "version": "0.1.0",
  "description": "An example Snap written in TypeScript.",
  "proposedName": "Transaction insights",
  "repository": {
    "type": "git",
    "url": "https://github.com/MetaMask/template-snap-monorepo.git"

>   },

Reviewer: ```suggestion
  },
  ...
}
```

Author: applied in latest commit

```

</TabItem>
<TabItem value="Full page" label="Example" default>

```json title="snap.manifest.json"
{
  "version": "0.1.0",
  "description": "An example Snap written in TypeScript.",
  "proposedName": "Transaction insights",
  "repository": {
    "type": "git",
    "url": "https://github.com/MetaMask/template-snap-monorepo.git"
  },
  "source": {
    "shasum": "9UBmDB7B0fFYh9tYoj8+NEN85MfqhkQQRc1MujXdq8w=",
    "location": {
      "npm": {
        "filePath": "dist/bundle.js",
        "packageName": "snap",
        "iconPath": "images/gas.svg",
        "registry": "https://registry.npmjs.org/"
      }
    }
  },
   "initialPermissions": {
    "snap_dialog": {},
    "endowment:rpc": {
      "dapps": true,
      "snaps": false
    },
    "endowment:network-access": {}
  },
  "manifestVersion": "0.1"
}
```
</TabItem>
</Tabs>

> #### 2.3. Update your Snap's button

Reviewer: I'm confused, can you clarify your reason for keeping this step? I find it confusing to ask users to update the text of a button that is non-functional.

We explain in the "Test your Snap" step that the button is non-functional, and in "Next steps" we suggest removing the button altogether.

If you prefer to keep something here about the button, we should guide users to remove the button instead of updating its text.

Author: We can probably prevent the div from being activated -- that wasn't in my wheelhouse: can investigate

Author: OK -- found a simple way to disable

    return <ConnectButton onClick={requestSnap} />;
  }

  if (shouldDisplayReconnectButton(installedSnap)) {
    return <ReconnectButton onClick={requestSnap} />;
  }

  return (
    <ConnectedContainer>
      <ConnectedIndicator />
      <ButtonText>Connected</ButtonText>
    </ConnectedContainer>
  );
};

````
</TabItem>
</Tabs>

These three updates are the minimum required to ensure that each user interaction with your Snap is well informed, however, your Snap will function without these tweaks.

### 3. Enable transaction insights and the Ethereum provider

To enable your Snap to provide [transaction insights](../../features/transaction-insights.md) and
use the global Ethereum provider, request the
[`endowment:transaction-insight`](../../reference/permissions.md#endowmenttransaction-insight) and
[`endowment:ethereum-provider`](../../reference/permissions.md#endowmentethereum-provider)
permissions in `packages/snap/snap.manifest.json`:

<Tabs>
  <TabItem value="Line updates" label="Update snap.manifest.json" default>

```json title="snap.manifest.json"
"initialPermissions": {
  "endowment:transaction-insight": {},
  "endowment:ethereum-provider": {}
}
```
</TabItem>

> <TabItem value="Full page" label="Example" default>

Reviewer: Similar to previous comments, I prefer omitting the full file tab.

Author: applied

A template test dapp displays, for installing and testing your Snap.

#### 5.2. Test your Snap

Select **Connect** and accept the permission request.

After connecting, you're prompted to install the Snap with the following permissions:
- **Access the Ethereum provider** 
- **Fetch and display transaction insights**

Next, select **Confirm** > **OK**.

From MetaMask Flask, create a new testnet ETH transfer. 

:::tip

You can set up [multiple accounts](https://support.metamask.io/managing-my-wallet/accounts-and-addresses/how-to-add-accounts-in-your-wallet/) to transfer between your accounts.

:::

Next, from the transaction confirmation window in MetaMask, switch to the tab named **TYPESCRIPT EXAMPLE SNAP** (or **Gas Percentage Calculator** if you followed Step 2).
Switching to the tab activates the [`onTransaction`](../../reference/entry-points.md#ontransaction)
entry point of your Snap and displays the percentage of gas fees in the transaction insights UI:

<p align="center">
<img src={require('../../assets/transaction-insights.png').default} alt="Transaction insights UI" width="400px" style={{border: '1px solid #DCDCDC'}} />
</p>

:::warning

> The activated **Send message** button is inoperable.

Reviewer: ```suggestion
The template **Send message** button is inoperable.
```

I think if you do Step 2, the button might not be "activated"?

> If you skipped Step 2, the activated **Send message** button is inoperable.

Author: ```suggestion
The activated **Send message** button is inoperable.
```

}
```

#### Customize your Snap

You can improve your Snap's UX by completing [Step 2](#2-optional-customize-your-snaps-ux). Consider 
updating `packages/site/src/pages/index.tsx` to remove, rather than updating the text of the 
non-functional template **Send message** button.

> Before publishing a Snap, it's also important to customize the metadata and properties of your app, 

Reviewer: ```suggestion
Before publishing a Snap, it's also important to customize the metadata and properties of your Snap, 
```

for example:

> - Update the `location` to its published location before publishing your Snap to `npm`, 

Reviewer: ```suggestion
- Update the `location` in `snap.manifest.json` to your Snap's published location.
- Update the `description` in `snap.manifest.json` to a description of your Snap.
```

</TabItem>
</Tabs>

#### Customize your Snap

You can improve your Snap's UX by completing [Step 2](#2-optional-customize-your-snaps-ux). Consider 
updating `packages/site/src/pages/index.tsx` to remove the 

> non-functional button.

Reviewer: ```suggestion
non-functional **Send message** button.
```

Sorry, add back.

```

:::tip
If you have previously developed a dapp, you're likely familiar with accessing the Ethereum provider 
using `window.ethereum`. In a Snap, the `window` object is not available. Instead, when you request 
the `endowment:ethereum-provider` permission, your Snap is granted access to the 
[`ethereum` global object](../about-snaps/apis.md#snap-requests).
:::

### 5. Build and test your Snap

Complete the following steps to build and test your Snap:

> #### 5.1. Increment version

Reviewer: Same as comment on the "Estimate gas" tutorial – this step might fit better under "Next steps," since it's not required for building/testing.

Author: agree, I didn't like it in the first pass I did, but thought it may have more purpose than I knew!
